### ISSUE TITLE
Can't use 'as' without first binding the variable to a name

### ISSUE THREAD

**Issue by Ruin0x11** [teal-language/tl#202]
This code throws a compiler error:

```lua
local Foo = record
   bar: boolean
end

local a: table = {}
(a as Foo).bar = true
```

```
/usr/bin/luajit: /usr/local/share/lua/5.1/tl.lua:2424: bad argument #1 to 'ipairs' (table expected, got nil)
stack traceback:
	[C]: in function 'ipairs'
	/usr/local/share/lua/5.1/tl.lua:2424: in function 'pretty_print_ast'
	stdin:1: in main chunk
	[C]: at 0x55ee50214c20
```

However, this code is fine.

```lua
local Foo = record
   bar: boolean
end

local a: table = {}
local b = a as Foo
b.bar = true
```

**Comment by hishamhm** [comment]
I can't reproduce the above, neither on `master` or 0.7.1:

```
local Foo = record
   bar: boolean
end

local a: table = {}
(a as Foo).bar = true
```

This gives me a syntax error (because of a parser ambiguity, which is arguably a bug). But if I add a semicolon this works as expected:

```
local Foo = record
   bar: boolean
end

local a: table = {};
(a as Foo).bar = true
```

Lua itself has a similar parser ambiguity, this is a syntax error:

```
local f = 0
local t = {}
local x = f
(t).hello = "world"
```

(The bug in Teal's parser is that the ambiguity is triggering for whenever the line starts with a `(` and the previous one ends with an expression, not only when it ends with something that could be a function call, such as an identifier or another parenthesized expression.)

**Comment by overcrook** [comment]
I recently ran into this problem too and did some research.

Let's take a look into this example



``` lua
local record Foo
   record Bar
      baz: boolean
   end
   bar: boolean
end

local a: table = {};
```

If we want to get a field `bar` from `a`, casted to `Foo`, we must write it as following:

```
(a as Foo).bar
```

If you omit the parentheses, you get a completely different expression:

```
a as Foo.bar
```

In this expression we cast `a` to type `Foo.bar`. Thus, parentheses are needed only for Teal itself to prioritize expressions.

Note that Teal translate an expression `a as Foo` as simple `a`. Parentheses is just another construct, and Teal do it's best keeping original semantics. But in this specific case we *really don't want* these parentheses.

So I made a hack. Mark somehow an output of a `operator as` and check for it while generating output for `paren`. If we find a cast inside paren, we remove unnecessary parens.

This looks and feels like a natural hack, and I am not satisfied by this solution, but it works for me. Please look at the attached diff and comment it. Maybe there can be an another, more elegant solution?

---
``` diff
diff --git a/tl.tl b/tl.tl
index 8ba616f..bf6f65e 100644
--- a/tl.tl
+++ b/tl.tl
@@ -2693,6 +2693,9 @@ function tl.pretty_print_ast(ast: Node, mode: boolean | PrettyPrintOpts): string
       {string}
       y: number
       h: number
+
+      -- HACK: explicit mark for output of 'operator as'
+      is_cast: boolean
    end
 
    local function increment_indent()
@@ -3082,9 +3085,17 @@ function tl.pretty_print_ast(ast: Node, mode: boolean | PrettyPrintOpts): string
       ["paren"] = {
          after = function(node: Node, children: {Output}): Output
             local out: Output = { y = node.y, h = 0 }
+
             table.insert(out, "(")
             add_child(out, children[1], "", indent)
-            table.insert(out, ")")
+
+            -- HACK: check if children[1] is an operator 'as' and remove parens in that case
+            if children[1].is_cast then
+                 -- remove the left paren
+                table.remove(out, 1)
+            else
+                table.insert(out, ")")
+            end
             return out
          end,
       },
@@ -3103,6 +3114,7 @@ function tl.pretty_print_ast(ast: Node, mode: boolean | PrettyPrintOpts): string
                table.insert(out, "]")
             elseif node.op.op == "as" then
                add_child(out, children[1], "", indent)
+               out.is_cast = true
             elseif node.op.op == "is" then
                table.insert(out, "type(")
                add_child(out, children[1], "", indent)

```

**Comment by hishamhm** [comment]
@overcrook Thanks for the feedback! However, I don't think this is the same problem.

`a as Foo.bar` meaning `a` as `Foo.bar` is the intended expansion, because Teal has support for nested record types. In that case `(a as Foo).bar` you really need to use parentheses.

**Comment by overcrook** [comment]
@hishamhm thanks, and yes, I know Teal supports nested record types and I tried to show that in my example (nested record `Bar` inside `Foo`). And I agree that I need to use parentheses to correctly cast `a` as `Foo`, not as `Foo.bar`.

But casting is a Teal concept only. In plain Lua I'd just write `a.bar`. The same thing, written in Teal as `(a as Foo).bar`, translates as `(a).bar`, and this is why I'm saying about unneeded parentheses *from Lua's point of view*.

Another thing is that Teal skips the semicolon at the end of the expression in the output. Let me show another example:

```
local record Foo
   bar: boolean
end

local a = {}
local b = a;
(a as Foo).bar = true
```

Note that I use a semicolon before the cast. So Teal is understands correctly that there are three statements. But in Teal's generated output I see that semicolon is missing, leading to an ambiguity for Lua.

```
local Foo = {}



local a = {}
local b = a
(a).bar = true
```
Maybe this is an another issue, but a combination of these produces a slightly different code than I expected to see.

I hope this information can be helpful for you.

**Comment by hishamhm** [comment]
@overcrook Ah, understood! Thanks for the patience explaining it. :) This case you described is a tricky code generation bug. I'm now wondering if anything other than casts may trigger it. If this affects casts only, might be that your hack above would be a usable interim solution for getting the correct behavior.

**Comment by hishamhm** [comment]
@overcrook @Ruin0x11 Happy to report that this is fixed since 0.11.0! (I forgot to close the issue earlier).

@overcrook's example above now compiles correctly (preserving the semicolon: b737364fe94bb375394a5b25f04d1712465e7e9e).
